### ISSUE TITLE
Support local builds

### ISSUE THREAD

**Issue by ndarilek** [boldpoker/edeliver#194]
### Environment

* Edeliver version (`mix edeliver --version`): N/A
* Elixir version (`elixir -v`): 1.4.0
* Build tool and version (distillery|exrm): Distillery
* Operating system (on build / deploy hosts): Linux
* Are you using an umbrella project (yes|no): No

### Current behavior

Edeliver expects that it has a remote build server.

### Expected behavior

If I don't specify a remote build server, I'd like for Edeliver to perform builds locally, perhaps with a warning/message stating that the build and deployment servers must be identical. I'm using [Concourse](https://concourse.ci), which performs builds by checking out code and bind-mounting it into Docker images. So, for instance, my code is checked out and built within the context of a Linux container whose setup is identical to that of my deployment server. The container isn't running an SSH server, nor is it remotely accessible. Docker is only used to set up an Elixir build environment which just happens to be identical to that on my staging/production VPS.

I'd like to use Edeliver to manage migrations, push/rollback releases, etc. I'd like it if it performed builds locally, so I can use this otherwise excellent tooling for everything but the build.

I've also considered just using the Distillery release process, then pointing its output location at where Edeliver expects its release store to be. Unfortunately there's no super obvious way to do this. Additionally, I'd really like to use Edeliver's easy auto-versioning. If there was an easy way to connect auto-versioning with Distillery's build process, use that for the build, then have the rest of the Edeliver suite just work, that'd be great too.

Also, I'm fairly new to Elixir, so if there's an obvious way to do this then please do let me know. :) I tried discussing this as a proposed feature on the forum but didn't get much in the way of responses, so am unclear as to whether or not that satisfied your issue requirements.

Thanks!

**Comment by eidge** [comment]
Hi there,

I have a similar use case, I'm using vagrant locally with a setup identical to the staging server and same thing happens with the CI setup so having a build server in my case seems like needless overhead/costs.

A local build option would definitely solve this for me as well.

Let me know if there's anything I can help with.

Thanks

**Comment by zhaohansprt** [comment]
I  also have a similar use case,   its very  inconvenient for test that can't do the local builds(failed on status 255),
 And beside that  the  production-deploy operation should be  separate out  being  able  to operate in different network  (i mean  with src build in a  private lan network and  with the sequenced output files that without src!!!    deploys in the internet network)

**Comment by DavidOliver** [comment]
I don't fully understand your desired process, but in case it helps you can have edeliver SSH into localhost and effectively build locally:

```
BUILD_HOST="localhost"
BUILD_USER="david"
BUILD_AT="/path/on/local/system"
```

I use the `asdf` language manager for Elixir, and due to differences in the local environment when SSH'ed into localhost I had to add this to `~/.profile` (the exact file may differ depending on distro):

```
# Source asdf so that 'mix' can be used when edeliver is SSH'ed into localhost.
if [ -f "$HOME/.asdf/asdf.sh" ] ; then
	. ~/.asdf/asdf.sh
fi
```

**Comment by erdosboti** [comment]
I have the exact same problem. It would be very nice to be able to build the release locally with exrm/distillery with all the niceties edeliver brings in to the process (like autoversioning) and then use edeliver to deploy that release from the local directory the release was created in.
@DavidOliver correct me if I'm wrong, but what you described means that you have to be able to ssh to localhost so you have to install an ssh-server and configure it. This seems to be additional complexity as we are able to create a release locally with either exrm or distillery. And since edeliver already uses these tools in the background it doesn't seem to be exceptionally hard to achieve (again correct me if I'm wrong).
In my concrete case I set up a docker image for development. Since this is the first time I'm trying to deploy a phoenix/elixir app I followed the guides and created a release with exrm easilly. Then I looked around and saw that edeliver could be used. But I don't have a build server (this is like a toy app) and I don't want to set up one. With current edeliver capabilities I only have 2 options:
- create a separate docker image for the build with ssh server in it
- modify my current docker image to also include ssh service
Both of them seem more complex than I'd expect if I'm already able to create a release.

**Comment by DavidOliver** [comment]
> @DavidOliver correct me if I'm wrong, but what you described means that you have to be able to ssh to localhost so you have to install an ssh-server and configure it. This seems to be additional complexity as we are able to create a release locally with either exrm or distillery.

Yes, I believe that's right. I'm not using Docker and configuring SSH was simple in my case.

**Comment by ndarilek** [comment]
There's basically nothing that hits the sweet spot of developing in a CI 
environment which matches your server, autoversioning, then pushing the 
finished build to an identical production server. The only way is to use 
Docker, or to make your CI server environment SSH-accessible.

I gave up on Edeliver because in the end it was easier to just use 
Docker, but it's odd to me that there isn't a Distillery++. Either merge 
the autoversioning into Distillery, or create a Deliver addon that can 
push the resulting build, then let Edeliver be Distillery with 
autoversioning + the push addon. It was annoying to set up a 
Docker-based build process just because I wanted to work around the 
assumption that, in order to autoversion and push a result, I had to 
either set up a remote build server or run an SSH server, when I'd 
already solved on my own the problems this was meant to.

Anyhow, guess this can be closed if the expectation is that I'm supposed 
to SSH to my local server instead of just running the build without the 
SSH step. It just seems like you've built a more complex solution than 
is necessary rather than letting me opt into the whole solution bit-by-bit.
